pyproject.toml

```
@@ -49,6 +49,11 @@
 [tool.black]
 line-length = 100
 
+[tool.isort]
```

Missing config options that should be added:
```suggestion
[tool.isort]
default_section = THIRDPARTY
indent = 4
known_first_party = ansible_navigator, key_value_store
known_testing = pytest, unittest
lines_after_imports = 2  # Ensures consistency for cases when there's variable vs function/class definitions after imports
no_lines_before = LOCALFOLDER  # Keeps local imports bundled with first-party
sections = FUTURE, STDLIB, TESTING, THIRDPARTY, FIRSTPARTY, LOCALFOLDER
use_parentheses = true  # Instead of escaping EOL
```

---

I'll add these one at a time, per commit to see the changes.

---

Note that some are coupled (you cannot inject a section that is not declared, for example).

---

default_section = THIRDPARTY appears to be the default https://pycqa.github.io/isort/docs/configuration/options.html#default-section

---

indent = 4, makes no changes in repo, leaving it out until such time it is needed

---

use_parentheses = true , makes no changes in repo, leaving it out until such time it is needed

---

known_first_party = "ansible_navigator, key_value_store", makes no changes in repo, leaving out

---

> known_first_party = "ansible_navigator, key_value_store", makes no changes in repo, leaving out

Please keep it. Probably there's no cases of confusion with third-party imports which may still happen.

---

It would appear isort is correctly determining that `ansible_navigator` is a first party though.... and the probability of the kv store causing an issue should be really low because of that directories use-case.

I'm happy to add it back if you feel it is important, but I think we should note that at the time of implementation it had no effect and provided no utility

---

Do you have a plan for automatically noticing when it stops being aligned? Formatters tend to make people ignore such problems so no hope of noticing them during manual human reviews.

---

There's also value in documenting the expectations transparently. Besides, when you integrate new linters, there may be no violations of the rules they bring but it's usually not a good reason for not linting things, right?

---

For the sake of history, can you provide an example of when in the future this might become necessary while I add it back?

---

added it:

```
known_first_party = "ansible_navigator, key_value_store" # No effect at implementation, here anticipating future benefit

```

---

`known_tests` has been removed, here's an example:

with: 
```
import os
import shlex

import pytest

from unittest import mock
```

without
```
import os
import shlex

from unittest import mock

import pytest
```

with this setting `pylint` throws: 
```
************* Module tests.unit.test_cli
tests/unit/test_cli.py:10:0: C0411: standard import "from unittest.mock import patch" should be placed before "import pytest" (wrong-import-order)
************* Module tests.unit.configuration_subsystem.test_configurator
tests/unit/configuration_subsystem/test_configurator.py:7:0: C0411: standard import "from unittest import mock" should be placed before "import pytest" (wrong-import-order)
tests/unit/configuration_subsystem/test_configurator.py:8:0: C0411: standard import "from unittest.mock import patch" should be placed before "import pytest" (wrong-import-order)
```
It's being removed to avoid the need to modify pylint.

It can be added at such time it is necessary in the future.

@webknjaz thanks for the suggestion here, it was a good test.

---

I suppose we just need to disable this rule in pylint. I've found an old issue that I've apparently created a while back that has the explanation that pylint uses isort under the hood: https://github.com/PyCQA/pylint/issues/3817#issuecomment-687892090.

I still think that having all test frameworks/libs grouped is benefitial.

---

> For the sake of history, can you provide an example of when in the future this might become necessary while I add it back?

I think that one of the major benefits is reproducibility as isort implementation changes and also the mere fact that it is helpful as documentation of "we expect that these modules/packages are local". Explicit is better than implicit: PEP 20 did not invent this principle out of nowhere. It's a good observation that when you want things to be readable/maintainable, presenting such information to people in an easily findable manner is polite and transparent. Otherwise, we'd be basically sending anybody who'd want to understand how/why it works like this to the side quest of trying to read the source code of isort — this is the price for having implicit undocumented assumptions that nobody knows about except for the implemented at the moment of initial addition.

---

@cidrblock look what I've found in the FAQ: https://pylint.pycqa.org/en/latest/faq.html#i-am-using-another-popular-linter-alongside-pylint-which-messages-should-i-disable-to-avoid-duplicates.

---

each has been vetted, and the final list is in the configuration, resolving conversation, please reopen if the concerns and question are unanswered, ty for the feedback. Some good ideas here and example of different config options.

---

In the last comments, I was asking to keep the testing section and disable the respective pylint rule per the official FAQ entry I've linked above.

---

but with the current implementation it is not necessary as there are no conflicts...

---

thinking aloud here, because this PR does not enforce `isort` disabling import-order in pylint now would create a period of time between this PR and a follow on PR where neither was enforce and could introduce additional errors in the repo.  I'm thinking it might be better to disable pylint import order checking is necessary and after isort via pre-commit is enforceing.

---

and in the case the team does not approve using isort as enforcing, changes to pylint will not need to be made or backed out

---

https://github.com/ansible/ansible-navigator/issues/784

---

https://github.com/ansible/ansible-navigator/issues/785